rust/lance-encoding/src/encodings/logical/dictionary.rs

```diff
@@ -0,0 +1,192 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright The Lance Authors
+
+use std::sync::Arc;
+
+use arrow_array::{
+    cast::AsArray,
+    types::{BinaryType, ByteArrayType, LargeBinaryType, LargeUtf8Type, UInt8Type, Utf8Type},
+    Array, ArrayRef, GenericByteArray, GenericListArray, LargeListArray, ListArray, UInt8Array, StringArray
+};
+
+use arrow_buffer::ScalarBuffer;
+use arrow_schema::{DataType, Field};
+use futures::future::BoxFuture;
+use lance_core::Result;
+use log::{debug, trace};
+
+use crate::{
+    decoder::{
+        DecodeArrayTask, LogicalPageDecoder, LogicalPageScheduler, NextDecodeTask, SchedulerContext,
+    },
+    encoder::{BatchEncoder, EncodeTask, FieldEncoder},
+};
+
+use super::{list::ListFieldEncoder, primitive::PrimitiveFieldEncoder};
+use std::collections::HashSet;
+
+use lance_arrow::deepcopy::deep_copy_array;
+
+
+#[derive(Debug)]
+pub struct AccumulationQueue {
```

Comment:
Can you reuse `primitive::AccumulationQueue`?

Reply:
We want to accumulate based on items, not bytes, so the new() and insert() function change (both parameters and body). Which is why I used a new queue. Do you still suggest reusing the old one?

Reply:
Maybe update the old one to support both approaches?  Although, now that I think about it, maybe accumulating based on bytes is a better idea anyways.  Otherwise, if we have a lot of very large strings, we might end up accumulating too much.